### ISSUE TITLE
Network bootstrap issues

### ISSUE THREAD

**Issue by runderwo** [gerbera/gerbera#903]
1. If an interface is not specified, Gerbera currently binds to an arbitrary IPv4 address, which can correspond to a virtual interface or even one with no active link.  This results in a frustrating silent failure.  Choosing 0.0.0.0 should be a better choice for IPv4, mirroring the :::49152 chosen for IPv6. 

2. Mediatomb previously added a static route for 239.0.0.0/8 to the bind interface that was chosen.  Gerbera is not doing this and as such the packets are not transmitted until this route is added manually.  An intelligent choice for the interface for this static route is probably related to the above choice since it's reasonable to expect upnp searches to be received on the same interface where the responses are to be emitted.

**Comment by whyman** [comment]
1. We dont really have any control over the bind behaviour, it is all handled by libupnp.

We don't want to bind to 0.0.0.0 if we can help it as we serve up the users files, and we don't want to inadvertently  end up somewhere public. 

2. Can you provide a case where the route would be needed? It works for me(tm)

**Comment by runderwo** [comment]
It's more accurate to say that we have some control over the bind behavior, as there's extensive policy around the interface supplied to libupnp: https://github.com/gerbera/gerbera/blob/c397cc511e6a2ea07848731e789617d7adf82538/src/server.cc#L100
I understand the concern about ending up somewhere public.  It's odd that libupnp would choose :: for ipv6 and not 0.0.0.0 for ipv4, isn't it?

The case where the static route would be needed is where there are multiple network interfaces on the host and the default route is the wrong interface for libupnp (e.g. WAN).

**Comment by KarlStraussberger** [comment]
I'm not sure whether MT really added it automatically because on openSuSE the startup exactly does this:
`ip route add 239.0.0.0/8 dev $MT_INTERFACE`
`ip link set dev $MT_INTERFACE allmulti on`
openSuSE also enforces the interface to be set.

I actually prefer having it set in my ifroute-file because it works regardless of MT or any other UPNP server is running.

**Comment by runderwo** [comment]
That's a good catch that it may have been done by startup scripts.

I wonder how we can:
1. do the right thing by default
2. if there's no right thing, how to avoid failing silently

As an engineer I was able to diagnose and fix the problem but I'm afraid a normal user might not have the same capacity.

**Comment by KarlStraussberger** [comment]
1. Might be a complicated thing because in most distros gerbera is not ran as root so we cannot at any network setting, So its a task for those who build the packages to remember network setup.
2. Is done Gerbera docs [Network Setup](https://docs.gerbera.io/en/latest/run.html#network-setup)

**Comment by runderwo** [comment]
Maybe it could be as simple as providing an "official" sample init.d and/or ExecStartPre script that does the needful things.

**Comment by whyman** [comment]
I suppose we can add it to the systemd unit we provide, its currently beautifully simple.

We would need some kind of env file or config file that you would need to edit to set the interface in addition to the gerbera config file which seems pretty gross from a users perspective.

We wont be running as root for the main process though, so we cant add the routes in from there.

Ive never run into a case where this has been needed though, its always just worked as expected.

**Comment by lyallp** [comment]
I have issues with Gerbera bootstrap on Gentoo Linux. 
Gerbera will start successfully after boot, once all the interfaces are on-line, etc.

During the boot process, Gentoo, or at least my system, brings IPv6 on-line before IPv4, with IPv4 arriving about 3 seconds later. 

As such, the startup script, which says Gerbera needs the net, thinks the net is on-line due to IPv6 availability and happily starts up Gerbera, which promptly crashes because IPv4 is not yet up.

Whilst not exactly the problem as described initially in this issue, I figure that should the connect to the IPv4 socket fail, maybe Gerbera could wait and retry, or, fail gracefully. The startup script on Gentoo can be configured to handle graceful failures and is happy with delays in startup but it gets messy when the app fails with a segfault.

Syslog of gerbera startup crash follows (lightly edited to remove superfluous entries and to fluff IPv6 addresses) :-

```
Jan  2 22:03:02 Lyalls-PC kernel: [   19.625318] IPv6: ADDRCONF(NETDEV_CHANGE): enp0s31f6: link becomes ready
Jan  2 22:03:02 Lyalls-PC dhcpcd[1897]: DUID 00:01:00:01:21:61:12:f7:70:4d:7b:67:83:60
Jan  2 22:03:02 Lyalls-PC dhcpcd[1897]: enp0s31f6: IAID 7b:67:83:60
Jan  2 22:03:02 Lyalls-PC dhcpcd[1897]: enp0s31f6: adding address fe80::16:cb74:9a0b:????
Jan  2 22:03:02 Lyalls-PC dhcpcd[1897]: enp0s31f6: rebinding lease of 192.168.1.9
Jan  2 22:03:02 Lyalls-PC dhcpcd[1897]: enp0s31f6: probing address 192.168.1.9/24
Jan  2 22:03:03 Lyalls-PC dhcpcd[1897]: enp0s31f6: soliciting an IPv6 router
Jan  2 22:03:03 Lyalls-PC dhcpcd[1897]: enp0s31f6: Router Advertisement from fe80::be05:43ff:fee6:????
Jan  2 22:03:03 Lyalls-PC dhcpcd[1897]: enp0s31f6: adding address 2001:44b8:24f:6200:4f72:303e:????:15dd/64
Jan  2 22:03:03 Lyalls-PC dhcpcd[1897]: enp0s31f6: adding route to 2001:44b8:24f:????::/64
Jan  2 22:03:03 Lyalls-PC dhcpcd[1897]: enp0s31f6: requesting DHCPv6 information
Jan  2 22:03:03 Lyalls-PC dhcpcd[1897]: enp0s31f6: adding default route via fe80::be05:43ff:fee6:????
Jan  2 22:03:06 Lyalls-PC kernel: [   23.133990] gerbera[2424]: segfault at 40 ip 000055d3be79414f sp 00007ffed7c5ffc0 error 4 in gerbera[55d3be6cc000+1bf000]
Jan  2 22:03:06 Lyalls-PC kernel: [   23.133993] Code: 66 ee 14 00 00 0f 84 e0 03 00 00 b8 ff ff ff ff f0 41 0f c1 44 24 08 83 f8 01 0f 84 e1 03 00 00 0f 1f 00 48 8b b5 70 01 00 00 <48> 8b 46 40 48 39 46 38 0f 84 73 01 00 00 31 db 4c 8d 7c 24 20 4c
Jan  2 22:03:06 Lyalls-PC /etc/init.d/gerbera[2423]: start-stop-daemon: /usr/bin/gerbera died
Jan  2 22:03:06 Lyalls-PC /etc/init.d/gerbera[2402]: ERROR: gerbera failed to start
Jan  2 22:03:07 Lyalls-PC ntpd[2266]: Listen normally on 6 enp0s31f6 [2001:44b8:24f:6200:4f72:303e:5f61:????]:123
Jan  2 22:03:08 Lyalls-PC dhcpcd[1897]: enp0s31f6: leased 192.168.1.9 for 864000 seconds
Jan  2 22:03:08 Lyalls-PC dhcpcd[1897]: enp0s31f6: adding route to 192.168.1.0/24
Jan  2 22:03:08 Lyalls-PC dhcpcd[1897]: enp0s31f6: adding default route via 192.168.1.254

```

**Comment by Nubicola** [comment]
I'm using Gerbera 1.1 on a raspberry pi (it's the latest version available). It'll take me some time to figure out how to cross-compile, so in the meantime I wonder if there's a workaround for this problem?

I edited the config.xml and specified to use eth0. I also tried putting the IP address in there (it's static), but on startup it's just ignored. The server thinks it's running on some random IP address. So far I've had to restart the server after boot is fully completed, then it gets the right IP address. If there's a workaround kindly let me know!

**Comment by lyallp** [comment]
A hack would be to put a 'sleep 5' in the start up script, let the network settle down before starting gerbera.
I am unfamiliar with your OS but maybe you have a 'delayed' start option or make Gerbera start after everything else?

**Comment by Nubicola** [comment]
> A hack would be to put a 'sleep 5' in the start up script, let the network settle down before starting gerbera. I am unfamiliar with your OS but maybe you have a 'delayed' start option or make Gerbera start after everything else?

Yes, this is the approach I took in essence. In detail, the system uses systemd and adding "After=network.target" for gerbera did the trick!
